### DISCUSSION TITLE
UAC2 microphone (under/over)run

### DISCUSSION THREAD

**Discussion by gmikitiuk** [hathach/tinyusb#2896] [General]
Hello,

I'm investigating microphone capabilities of UAC2 class and I'm not clearly understanding how microphone shall work without feedback endpoint and asynchronous endpoint - depending of clock accuracy of the USB device microphone samples might be produced at slightly higher/slower rate than USB host picks them up which eventually will lead to overrun/underrun.

At least that's my observation during tests with audacity where I explicitly simulate situation where I wouldn't have anything to put into the usb buffers.

What did I miss?

**Comment by HiFiPhile** [comment]
> how microphone shall work without feedback endpoint and asynchronous endpoint

You should use asynchronous endpoint in real case, otherwise your I2S microphone have to use the same USB clock.

Feedback is only useful for OUT transfer since the host needs to know how many samples it needs to send, for IN transfer it's the device itself who decides how many samples it will send to PC.

You can look at option `CFG_TUD_AUDIO_EP_IN_FLOW_CONTROL` and #2282, basically:
- Send nothing if FIFO count < n<sub>avg</sub>-1
- Send n<sub>avg</sub>-1 samples if FIFO depleting
- Send n<sub>avg</sub> samples if FIFO stable
- Send n<sub>avg</sub>+1 samples if FIFO increasing
